[RFCs FS-1051, FS-1052, FS-1053] support for span, readonly refs, byref-like structs

dsyme opened the pull request:
Adding support for some C# 7.0-7.2 features:

RFCs
* https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1053-span.md

Simple samples compile, e.g. taken from [here](https://github.com/dotnet/corefxlab/blob/master/docs/specs/span.md)
```fsharp
open FSharp.NativeInterop
open System.Runtime.InteropServices

let SafeSum (bytes: Span<byte>) =
    let mutable sum = 0
    for i in 0 .. bytes.Length - 1 do 
        sum <- sum + int bytes.[i]
    sum

let test() = 
    // managed memory
    let arrayMemory = Array.zeroCreate<byte>(100)
    let arraySpan = new Span<byte>(arrayMemory);
    SafeSum arraySpan |> printfn "res = %d"

    // native memory
    let nativeMemory = Marshal.AllocHGlobal(100);
    let nativeSpan = new Span<byte>(nativeMemory.ToPointer(), 100);
    SafeSum nativeSpan |> printfn "res = %d"
    Marshal.FreeHGlobal (nativeMemory);

    // stack memory
    let mem = NativePtr.stackalloc<byte>(100)
    let mem2 = mem |> NativePtr.toVoidPtr
    let stackSpan = Span<byte>(mem2, 100)
    SafeSum stackSpan |> printfn "res = %d"
```

TODO
* [x] factor out the cleanup to keep this minimal (and perhaps revert some of it)
* [x] test `ReadOnlyMemory` and `ReadOnlySpan`
* [x] design: byref returns should automatically dereference unless `&` is used. currently you must assign to a local and then use that local (where the use of the local does an automatic dereference)
* [x] resolve all unresolved questions in RFC
* [x] add positive tests
* [x] add negative tests
* [x] Fix this baseline https://github.com/Microsoft/visualfsharp/pull/4888/files#diff-24bfe33fd52123c01906c88689b77404R2
* [x] Add positive tests for emit of `modreq` on `inref` parameters
* [x] Consider [design issue re stackalloc](https://github.com/Microsoft/visualfsharp/pull/4888/#issuecomment-393624628)
* [x] Fix [bug in subsumption calls](https://github.com/Microsoft/visualfsharp/pull/4888/#issuecomment-393628869)
* [x] Improve an [error message](https://github.com/Microsoft/visualfsharp/pull/4888/#issuecomment-393640604)
* [x] Fix  [assignment to return byref](https://github.com/Microsoft/visualfsharp/pull/4888/#issuecomment-393648057)
* [x] Fix  [assignment to byref using setter syntax](https://github.com/Microsoft/visualfsharp/pull/4888/#issuecomment-393648057)
* [x] Fix [immediate return of span local](https://github.com/Microsoft/visualfsharp/pull/4888#issuecomment-393993472)
* [x] Check for both IsByRefLike and Struct
* [x] Review rules on Span safety and differentiate between byref and span safety rules, disallowing byref fields in IsByRefLike types.

dsyme commented:
Right now proto is broken, am fixing that

Once this starts to work I'd greatly appreciate help in testing/proofing it. It's not a feature I currently need so I'm likely to miss things. We missed at least one thing with ref returns (mentioned above - we should just fix it since ref returns are currently very rare) and that was because it was not proofed enough

cartermp commented:
cc @drvink and @zpodlovics who should have some use cases to throw at this

dsyme commented:
@cartermp I think it's a matter of systematically translating C# tests and samples (as I did above).

smoothdeveloper commented:
@dsyme @cartermp "Future of C#" video from Build 2018 has a section about C# refs:

https://youtu.be/QZ0rWLaMZeI?t=1214 

# using `this ref` as extension member

Consider making F# able to call C# "`this ref` extension methods", and possibly to define those as well.

# ref return / ref locals

* A local variable can be a shared reference to another field / variable
* ref readonly return / ref readonly locals, not sure how the semantics will play out with our dealing with mutable/immutable bindings, I expect F# to at least be able to define ref readonly return methods
* escape analysis of taking a ref of something which goes out of scope

C# also added a new modifier beside `out` and `ref`: `in`, it acts the same as `ref` but ensure that the callee can't reassign the ref to a new value.

dsyme commented:
@smoothdeveloper Thanks, yes, I've added those to the list of TODO items. Some of it is already done as part of F# 4.1 ref returns but there are more details to do. This series https://blogs.msdn.microsoft.com/mazhou/tag/c/ is a good guide

zpodlovics commented:
@dsyme Please note ref local reassignment is now allowed in C#:

https://github.com/dotnet/csharplang/issues/933
https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.3/ref-local-reassignment.md
https://github.com/dotnet/csharplang/blob/master/meetings/2017/LDM-2017-10-02.md

vasily-kirichenko commented:
@dsyme the first two links to RFCs does not work.

saul commented on src/fsharp/FSharp.Core/nativeptr.fsi:
```diff
@@ -19,6 +19,22 @@ namespace Microsoft.FSharp.NativeInterop
+        [<CompiledName("ToVoidPtrInlined")>]
+        /// <summary>Returns a typed native pointer for a given machine address.</summary>
+        /// <param name="address">The pointer address.</param>
+        /// <returns>A typed pointer.</returns>
```
Return value is untyped

saul commented on src/fsharp/FSharp.Core/nativeptr.fsi:
```diff
@@ -19,6 +19,22 @@ namespace Microsoft.FSharp.NativeInterop
+        [<Unverifiable>]
+        [<NoDynamicInvocation>]
+        [<CompiledName("ToVoidPtrInlined")>]
+        /// <summary>Returns a typed native pointer for a given machine address.</summary>
```
"'Returns an untyped"

dsyme commented:
I did a fair bit of work on this today and updated the RFC and implementation accordingly.

The biggest addition was in the way we handle in/out pointers and type inference issues associated with these.  The implementation became smooth and simple immediately we switched to the mechanism described briefly in https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1053-span.md#add-byreft-kind-to-represent-pointers-with-different-capabilities

dsyme commented:
Added 
* byref extension methods 
* implicit address-of  `inref` arguments at callsites
* fixes
* more tests

dsyme commented:
@cartermp This work is basically code complete (need to get it green of course, but that's mostly just updating baselines now).  

The byref tests are running but the span tests are only compiling, not executing, since I can't guarantee to have .NET 4.7.2.

dsyme commented:
Yay, green!

dsyme commented:
@dotnet-bot build Ubuntu16.04 Release_fcs Build please

dsyme commented:
@dotnet-bot test Ubuntu16.04 Release_fcs Build please

dsyme commented:
I've been using an extended Span sample on .NET 4.7.2 and noticed a problem where we can't access the `AsMemory` extension method, e.g.:

```
            let arrayMemory = Array.zeroCreate<byte>(100)
            let arraySpan = new Span<byte>(arrayMemory);
            arraySpan.AsMemory() 

The array version of the extension method seems to always be taking priority.

This is the only current issue I know of.

dsyme commented:
> I've been using an extended Span sample on .NET 4.7.2 and noticed a problem where we can't access the AsMemory extension method

I was wrong about this, the extension methods are only available on the array and ArraySegment types

dsyme commented:
Noticed when processing feedback from initial design review by @jaredpar and @OmarTawfik 

* [ ] Fix this baseline https://github.com/Microsoft/visualfsharp/pull/4888/files#diff-24bfe33fd52123c01906c88689b77404R2

* [ ] Review other negative tests, e.g. add negative test for `IsReadOnly` on mutable struct

* [ ] Add positive tests for emit of `modreq` on `inref` parameters

dsyme commented:
@dotnet-bot test Ubuntu16.04 Release_fcs Build please

dsyme commented:
This is ready, but needs both close review and active use...

TIHan commented:
I tested this on one my personal projects where I used Span and ref returns a fair amount. Here is what I found so far, I will write up more as I go along:

Original code:
```fsharp
    let (<--) (r: byref<byte>) (v: byte) =
        r <- v

    let write8 (data: Span<byte>) offset value =
        data.[offset] <-- byte value
```

```data.[offset] <-- byte value``` does not compile anymore due to de-referencing according to the spec of ref returns, so I had change it to this:
```fsharp
&data.[offset] <-- byte value
```

Not a big deal as the spec makes sense to me. However, do realize that this change to ref returns is a breaking change. This means existing code like this:
```fsharp
let ec = entComps.GetByRef(entity.Index)
ec.AddComponentId(uint16 bit)
```
will technically still compile, but behave differently. Before, `ec` was of type, `byref<EntityComponents>`. Now it is of type, `EntityComponents`, due to de-referencing. I will not receive any compile errors or runtime errors on this, the behavior of the program has now changed and I need to go fix it by adding the `&`.

Also, I would have expected this to work:
```fsharp
    let write8 (data: Span<byte>) offset value =
        data.[offset] <- byte value
```
This is what C# allows, assigning values to byrefs returned from functions. In this case, `Item`. Basically, they shouldn't deference if they are being assigned.

TIHan commented:
I found another issue, specifically with Span:
```fsharp
let test () =
        let alloc = NativePtr.stackalloc<byte> 10 |> NativePtr.toNativeInt
        Span<byte>(alloc.ToPointer(), 10) 
```

This compiles, but this doesn't:
```fsharp
let test () =
        let alloc = NativePtr.stackalloc<byte> 10 |> NativePtr.toNativeInt
        let s = Span<byte>(alloc.ToPointer(), 10) 
        s
```
Basically, the first snippet shouldn't compile.

TIHan commented:
Found another bug:
```fsharp
type Doot() = class end

[<Sealed>]
type GiantDad() =

    let test (data: Span<byte>) (doot: Doot) =
        ()

    member __.Test(data: Span<byte>) =
        test data Unchecked.defaultof<Doot>
```
This doesn't compile when it should.

TIHan commented:
This doesn't compile, which it shouldn't:
```fsharp
type Beef() =
    
    member this.Create() =
        let mutable bef = Unchecked.defaultof<beef>
        this.Create(&bef)

    member this.Create(bef: byref<beef>) =
        Span.Empty
```
But we want to be sure the corresponding C# equivalent doesn't either. Also the error message may not be clear:
```The address of the variable 'bef' cannot be used at this point. A method or function may not return the address of this local value.```

dsyme commented:
Thanks @TIHan that's great.  I've listed these as 5 items at the top of the PR and noted the design-related ones in the RFC.  

They are all pretty easy to work through, I'll aim to do them now

dsyme commented:
@cartermp @TIHan @jaredpar Thanks for the very extensive testing and review, that has been incredibly useful.

I've updated the RFC with resolution of all outstanding design issues here: https://github.com/fsharp/fslang-design/pull/308/

I've updated the implementation to correspond, specifically
* [x] Fix bug in subsumption calls
* [x] Improve an error message
* [x] Fix assignment to return byref `C.M() <- expr`
* [x] Fix assignment to byref using setter syntax `c.[i] <- expr`
* [x] Implicit deref of byref returns for let-bound functions

Some baselines will need to be updated, I'll do those based on the CI runs in the morning here

thanks
don

TIHan commented:
No problem! I will go ahead and test these news changes along with .net core 2.1.

dsyme commented:
@TIHan super, thanks

The build appears to be mostly-green (baselines will need updating) so I think I'll head to bed and see what you've found in the morning :)

TIHan commented:
I haven't managed to compile anything with the VS hive because it looks like it is not using the deployed compiler for some reason. Tomorrow, I will look into why.

But just from changing a few things in my personal project and seeing the editor giving me feedback, it seems to be fine and as intended. Also, .NET 2.1 seems to be fine as well, though I'm not sure what we are looking for here.

I did notice this though:
```fsharp
type Beef() =

    let rec beef id (data: byref<byte>) =
        beef (id + 1) &data
```
So, apparently that isn't legal, but this is:
```fsharp
module Beef =

    let rec beef id (data: byref<byte>) =
        beef (id + 1) &data
```
Is there a specific reason for that? (this also happens without span/ref changes)

dsyme commented:
@dotnet-bot Test Ubuntu16.04 Release_default Build please

dsyme commented:
> I did notice this though... 

@TIHan Thanks, I've addressed that case and added a lot more tests in that area.

dsyme commented:
OK, all green now!

smoothdeveloper commented on tests/fsharp/tests.fs:
```diff
@@ -1,5 +1,5 @@
-﻿#if INTERACTIVE
-//#r @"../../release/net40/bin/FSharp.Compiler.dll"
+﻿// vvvvvvvvvvvvv To run these tests in F# Interactive , 'build net40', then send this chunk, then evaluate body of a test vvvvvvvvvvvvvvvv
```
> vvvvvvvvvvvvv 

is this the same as "visual studio is lagging bad?" 😄

TIHan commented:
Found another issue, this should compile:
```fsharp
let test () =
    let s = Span<byte>.Empty
    s
```
This is somewhat related to the stackalloc problem, but we are ignoring checks for stackalloc for now. The other should compile, just like C#:
```csharp
        public Span<byte> Test()
        {
            var s = Span<byte>.Empty;
            return s;
        }
```

TIHan commented:
Here is another example similar to above comment:
```fsharp
let beefy = [|0|]
let testByref_this_doesn't_compile() =
    let x = &beefy.[0]
    &x

let testByref_this_compiles() =
    &beefy.[0]
```

TIHan commented:
Also, this code:
```fsharp
type Jopac() =

    member this.Create() =
        let mutable x = 1
        this.Create(&x)

    member __.Create(x: byref<int>) =
        Span<int>.Empty
```
Doesn't compile, as I remember @dsyme and I talking that this is to be extra careful. However, in C#, you can do this:
```csharp
        public Span<byte> Test(ref int data)
        {
            return Span<byte>.Empty;
        }

        public void CallingStuff()
        {
            var x = 1;
            Test(ref x);
        }
```
So now I question if we should allow it compile.


Some other examples:
```fsharp
type Jopac_NotCompile_WhichIsCorrect() =

    member __.Create(x: byref<int>) =
        &x

    member this.Create() =
        let mutable x = 1
        let x = &this.Create(&x)
        &x

    member this.CreateAgain() =
        let mutable x = 1
        &this.Create(&x)

type Jopac_NotCompile_WhichIsMightBeIncorrect() =
    
    member __.Create(x: byref<int>) =
        Span<int>.Empty

    member this.Create() =
        let mutable x = 1
        let x = this.Create(&x)
        x

    member this.CreateAgain() =
        let mutable x = 1
        this.Create(&x)
```

This C# code compiles:
```csharp
        public Span<int> Test(ref int x)
        {
            return Span<int>.Empty;
        }

        public Span<int> CallingStuff()
        {
            var x = 42;
            return Test(ref x);
        }
```

But this C# code doesn't compile:

```csharp
        public ref int Test2(ref int x)
        {
            return ref x;
        }

        public ref int CallingStuff2()
        {
            var x = 42;
            return ref Test2(ref x);
        }
```

@jaredpar maybe you can explain what might be going on here? F# is treating Span as if it is a byref, essentially, which we are calling "byref-like".

jaredpar commented:
@TIHan 

> This C# code compiles:

This is legal but it's hard to understand why by just reading the [Span safety spec](https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md) as there are some missing details. The key detail is that the `Span<T>` type cannot, by explicit design, turn `ref` parameter into a length one `Span<T>` over the location. 

``` csharp
void M(ref int x) { 
  // No such constructor available
  var s = new Span<int>(ref x);
}
```

This also by implication means there is no `ref` struct out there which can do this either. There is no such thing as a `ref` field in C#, or .NET really. Hence if `Span<T>` can't capture a `ref` in this way then no other `ref struct` type can either without unsafe code (for which verification isn't interesting). 

Another, possibly less confusing, way to think about this is the following:

> It is not possible to have a `ref` field in .NET nor is it possible to approximate one by using a `Span<T>`. 

Given that rule it's easier to see why the `CallingStuff / Test` interaction is safe. 

Note: I will be updating our spec to call out this case as it's important to understanding how our rules hold together. 

> But this C# code doesn't compile:

Correct. This is covered in the Method Invocation section:

> An lvalue resulting from a ref-returning method invocation e1.M(e2, ...) is ref-safe-to-escape the smallest of the following scopes:

Here the smallest scope is `ref x` which is the scope of the method body hence the return is not safe to return.

TIHan commented:
Thank you for your explanation @jaredpar .

TIHan commented:
@dsyme , for `[<IsByRefLike>]` attribute, I'm trying to compile this code that was mentioned in the RFC:
```fsharp

[<IsByRefLike>]
type S(count1: byref<int>, count2: byref<int>) = 
    member x.Count1 = count1
    member x.Count2 = count2
```

And it doesn't compile.

Also, I think there are some problems with this. `IsByRefLike` should only work on structs. In the example above, it's trying to do it on a class. We should require both `[<IsByRefLike;Struct>]`. Currently, if we try to do that it will still not compile as it says its illegal, same for `[<IsReadOnly>]`. Also, we should never allow byref fields, however it is fine to have byref-like structs as fields for byref-like types. @jaredpar , can you correct me if I'm wrong about thatbeef

TIHan commented:
Also @beefsyme, I'm trying to start writing some more examples based on these: https://github.com/dotnet/csharplang/blob/master/proposals/csharp-7.2/span-safety.md#examples

dsyme commented:
Thanks, yes, I forgot to track this one:
```
let test () =
    let s = Span<byte>.Empty
    s
```
Have added it, will fix it in the morning. Have noted it.  

> @dsyme , for [<IsByRefLike>] attribute, I'm trying to compile this code that was mentioned in the RFC

Yes, the Struct attribute is needed.  We can add a check for that.

> This is legal but it's hard to understand why by just reading the Span safety spec as there are some missing details. The key detail is that the Span<T> type cannot, by explicit design, turn ref parameter into a length one Span<T> over the location.

I'll review the Span safety spec.  By the look of it we're being conservative, and limiting code along the same lines as byref. I would be comfortable with a very slightly limited version of this feature in F# 4.5 if there are some cases like this, since the feature meets its core goal of being able to consume and use Span and byref-return APIs.

Please double check all known issues are tracked by tickbox items at the top, feel free to edit them in.

dsyme commented:
> Also, we should never allow byref fields, however it is fine to have byref-like structs as fields for byref-like types. @jaredpar , can you correct me if I'm wrong about that?

Currently the intent of the RFC is to  feature-limit this in F# 4.5, and not allow the declaration on new IsByRefLike structs with byref fields.  Perhaps it is harmless and easy, simply allowing the declarations.  But I think it would be ok to feature limit this in F# 4.5 as again, the RFC and PR meets to core goal of consuming and using Span/Memory APIs.  Let me know - I mean, we can allow it, but we just don't have to.

TIHan commented:
I think it is fine to feature limit IsByRefLike and IsReadOnly struct creation as I don't see an immediate need for it. As long as we track that is what we do for future versions.

Though, we should definitely allow calls to functions that take a byref or a span that return a span.

zpodlovics commented:
Please do not limit the usefullness of IsByRefLike struct creations, at least allow it with an experimental attribute (that may change in the future) or something like that. Some of the developers may only interested to consume and use Span and byreflike  from C# but **I would like to also able to produce it.** 

What the marketing should put about F# to the marketing brossure ? F# introduced Span support, but I we still have to write the core libraries in C# because F# Span support is unusable for Span production...

dsyme commented:
@zpodlovics The RFC is very clear: 

> The main aim of the RFC is to allow effective consumption of APIs using Span, Memory and ref-returns.

To be constructive, please contribute examples of new `IsByRefLike` types that have `Span` as fields, where you couldn't just pass 2-3 of `Span` arguments instead (if you had any more `Span` values then the performance costs of copying would likely go through the roof).  I actually doubt you'll find any truly compelling examples, but it would be great to have them if you do.

zpodlovics commented:
@dsyme I can do interlocked/atomic operations with a single  `IsByRefLike` type safely because it works with reference (byref) like semantics, but I cannot do that with passing 2-3 as fields / arguments / around.

For example, I have an in-memory shared lock-free data structure and multiple threads and multiple processes access the same data structure, as a simple producer / consumer model:

```csharp
    ref struct DataView<T>
    {
        public Int64 version;
        public Span<T> data;
    }
```

Both the producer both the consumer able to access both the data memory region both the control memory region (eg.: producer -> consumer notifications), and they do their work in a lock-free way. When the data organization is log structured too (append only), these data could be also processed in a really nice functional way.

And I would like to do both produce and consume these records concurrently in F# preferrably in a nice functional way, because it's far easier to compose them. There is no real alternative here as allocation is prohibitive (10M+ / sec) only the fully mutable objects, and object pooling makes it usable.

What is the intention to prevent IsByRefLike and IsReadOnly usage? As described in other issue Span is never indended for the regular users/developers, but only the developers who works on performance critical codes. Why a compiler warning is not enough something like what F# already have for inline IL? If something could goes wrong in F# definitely inline IL is a perfect example. 

And yes, something may goes wrong when somebody works with these performance critical codes - but this is expected, as any real concurrency related work will not be easy. But preventing these performance related works may also means loss of opportunity and loss of new functionality in the future, for example there would be no ``NativePtr.toByRef`` functionality (also marked for F# 4.5 target) with disabled inline IL support:

https://github.com/Microsoft/visualfsharp/issues/409
https://github.com/Microsoft/visualfsharp/pull/3691

dsyme commented:
Re this:

```
    ref struct DataView<T>
    {
        public Int64 version;
        public Span<T> data;
    }
```

OK, so if I understand correctly you want to pass a _byref_ to one of these values down the stack?  Because if you pass the value itself it is non-atomic.

I'll assess this with @TIHan

TIHan commented:
Ok, so this should compile according to the C# examples:

```fsharp
type SpanLikeType = Span<int>

let m1 (x: byref<Span<int>>) (y: Span<byte>) =
    // this is all valid, unconcerned with stack-referring stuff
    let local = SpanLikeType()
    x <- local
    x
```

I can't do a type alias to Span<int>, but I don't know if that makes sense to do.

TIHan commented:
Ok, here are the samples from C#:
```fsharp
type SpanLikeType = Span<int>

let m1 (x: byref<Span<int>>) (y: Span<byte>) =
    // this is all valid, unconcerned with stack-referring stuff
    let local = SpanLikeType()
    x <- local
    x

let test1 (param1: byref<Span<int>>) (param2: Span<byte>) =
    let stackReferring1 = Span<byte>()
    let mutable stackReferring2 = Span<int>()

    // this is allowed
    stackReferring2 <- m1 &stackReferring2 stackReferring1

    // this is NOT allowed
    stackReferring2 <- m1 &param1 stackReferring1

    // this is NOT allowed
    param1 <- m1 &stackReferring2 stackReferring1

    // this is NOT allowed
    let param2 = stackReferring1.Slice(10)

    // this is allowed
    param1 <- Span<int>()

    // this is allowed
    stackReferring2 <- param1

let m2 (x: byref<Span<int>>) =
    // should compile
    &x

let test2 (param1: byref<Span<int>>) (param2: Span<byte>) =
    let stackReferring1 = Span<byte>()
    let mutable stackReferring2 = Span<int>()

    let stackReferring3 = &(m2 &stackReferring2)

    // this is allowed
    stackReferring3 <- m1 &stackReferring2 stackReferring1

    // this is allowed
    (m2 &stackReferring3) <- stackReferring2

    // this is NOT allowed
    (m1 &param1) <- stackReferring2

    // this is NOT allowed
    param1 <- stackReferring3

    // this is NOT allowed
    &stackReferring3

    // this is allowed
    //&param1 - uncomment to test return
```

Currently, we violate some of these. Hopefully I did the conversion correctly here.

TIHan commented:
I want to mention these stacks rely on stack information, so this code:
```fsharp
    let stackReferring1 = Span<byte>()
    let mutable stackReferring2 = Span<int>()
```
Isn't accurate, we need to find something analogous to this:
```csharp
   Span<byte> stackReferring1 = stackalloc byte[10];
   var stackReferring2 = new SpanLikeType(stackReferring1);
```

dsyme commented:
OK, @TIHan and I have been over all remaining known cases and fixes and tests are now in for all of them.

TIHan commented:
test Windows_NT Release_ci_part2 Build please

TIHan commented:
Currently, tests are failing on the `core/byrefs/test.fsx`, I believe it is due to this code:
```fsharp
        let test() = 
            let addr = &f (C()) 
            addr <- addr + 1
            let addr = &f (ObjExpr()) 
            addr <- addr + 1
            check2 "cepojcwem16" 3 x
```
It's failing on this line `let addr = &f (C())` with:
*
Error	FS0412	A type instantiation involves a byref type. This is not permitted by the rules of Common IL.
*
And there are two of them.

TIHan commented:
Another minor issue:
```fsharp
[<Struct>]
type TestMut =

    val mutable x : int

    static member Test(x: inref<TestMut>) = ()

    static member CallTest() =
        TestMut.Test(TestMut())

let testIn (m: inref<TestMut>) =
    ()

let callTestIn() =
    testIn (TestMut())
```
This doesn't compile due to this line: `testIn (TestMut())`. It looks like let-bound functions are not using the same rules as member functions when instantiating something inline when calling a function that takes a `inref`. This could be a design decision for let-bound functions, but I think they should be the same as member functions. We should have consistency for refs across let-bound and member functions, otherwise it would be confusing.

TIHan commented:
This is another minor issue:
```fsharp
[<Struct>]
type TestS =

    static member Test(x: outref<TestS>) = ()

let testIn (m: outref<TestS>) =
    ()
```
This should not compile and give you an error saying that your `outref` variable has not been assigned. This means we should be checking to make sure the `outref` variable gets assigned in all branches of the body of the function. Technically, we were able to do `[<Out>]` on a parameter before, but it never did do any checks as well.

Since the checks haven't been there before anyway, I don't think we have to include this check in 4.5 - we could do it later if it requires a lot more work.

dsyme commented:
Thanks. All fixes are in and being validated in CI.


> This could be a design decision for let-bound functions, 

Yes, it is by design. I've noted it in the RFC.  I understand why you would expect the same conversions to take place, but we currently don't do any such conversions (e.g. F# lambda --> System.Func) on calls to let-bound functions, and I'm loathe to start doing them here , it would be better to make a systematic change here that took into account all conversions.

> ...out writing... Since the checks haven't been there before anyway, I don't think we have to include them in 4.5 - we could do them later if it requires a lot more work.

Yes, we don't make this check today and it won't be part of F# 4.5. I've noted that in the RFC

> But, looking at the JIT ASM for either, they generate the same ASM code. So it probably does not matter, just wanted to make note of it.

Yes, it's ok, we generate some more locals but the JIT does a good job of them

TIHan commented:
test Ubuntu16.04 Release_default Build please

dsyme commented:
@dotnet-bot test Ubuntu16.04 Release_default Build please

TIHan commented:
And we got it! Any bugs we find, we will thoroughly review with the C# team; but so far, I feel this is ready considering the massive testing we have done, especially for a personal project that uses this stuff heavily.
